app/javascript/components/visualization/ScatterPlot.jsx

```diff
@@ -371,90 +375,60 @@ function getPlotlyTraces({
     scatterColor: dataScatterColor,
     annotParams: { name: annotName, type: annotType },
     customColors = {}
-  }
+  },
+  activeTraceLabel,
+  expressionFilter
 }) {
-  const trace = {
+  const unfilteredTrace = {
     type: is3D ? 'scatter3d' : 'scattergl',
     mode: 'markers',
     x: data.x,
     y: data.y,
     annotations: data.annotations,
+    expression: data.expression,
     cells: data.cells,
     opacity: pointAlpha ? pointAlpha : 1
   }
   if (is3D) {
-    trace.z = data.z
+    unfilteredTrace.z = data.z
   }
 
-  let countsByLabel = null
 
   const isRefGroup = getIsRefGroup(annotType, genes, isCorrelatedScatter)
+  const isGeneExpressionForColor = !isCorrelatedScatter && !isAnnotatedScatter && genes.length
+
+  const startTime = performance.now()
+  const [traces, countsByLabel, expRange] = filterTrace({
+    trace: unfilteredTrace,
+    hiddenTraces, groupByAnnotation: isRefGroup, activeTraceLabel, expressionFilter, expressionData: data.expression
+  })
+  const perfTime = performance.now() - startTime
```

Comment:
it would be cool --but I don't think it's worth the added code complexity to surface it given that this is currently the fastest part of our visualization codebase,  and the most easily simulated in development.   We're already starting to see user-impacting bugs due to the extra complexity of our analytics codebase -- e.g. the study bulk download being broken because a nested analytics method was being called with unexpected parameters, so I'm leery of altering function signatures for analytics unless we have a clear need for it